### PR TITLE
[DomCrawler] Fix getting text after a child node with Crawler::innerText()

### LINKED ISSUE #48682: DomCrawler component Crawler::innerText() doesn't get text after a child node
### Symfony version(s) affected

5.4

### Description

The test case for this method basically is: 

```html
<div id="complex-element">
    Parent text
    <span>Child text</span>
</div>
```

Getting `$crawler->text()` then returns `Parent text Child text` and `$crawler->innerText()` returns `Parent text`. So far, so good.

If I don't misunderstand anything, in case of:

```html
<div id="complex-element">
    Parent text
    <span>Child text</span>
    More parent text
</div>
```

the return value of `$crawler->innerText()` should be `Parent text More parent text`. But it's still just `Parent text`.

I'll provide a pull request to fix this in a minute.

### How to reproduce

In the `Tests/AbstractCrawlerTest.php` test file, adapt the HTML in the `createTestCrawler()` method from: 

```html
<div id="complex-element">
    Parent text
    <span>Child text</span>
</div>
```

to

```html
<div id="complex-element">
    Parent text
    <span>Child text</span>
    More parent text
</div>
```

And the `testInnerText()` method accordingly to:
```php
public function testInnerText()
{
    self::assertCount(1, $crawler = $this->createTestCrawler()->filterXPath('//*[@id="complex-element"]'));

    self::assertSame('Parent text Child text More parent text', $crawler->text());
    self::assertSame('Parent text More parent text', $crawler->innerText());
}
```

### Possible Solution

https://github.com/symfony/symfony/pull/48684 

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by otsch**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #48682
| License       | MIT

### CODE REVIEW

File: src/Symfony/Component/DomCrawler/Crawler.php [hunk 1]
```diff
@@ -623,18 +623,39 @@ public function text(string $default = null, bool $normalizeWhitespace = true)
...
     /**
      * Returns only the inner text that is the direct descendent of the current node, excluding any child nodes.
+     *
+     * @param string|null $default             When not null: the value to return when the current node is empty
+     * @param bool        $normalizeWhitespace Whether whitespaces should be trimmed and normalized to single spaces
+     *
+     * @throws \InvalidArgumentException When current node is empty
      */
-    public function innerText(): string
+    public function innerText(string $default = null, bool $normalizeWhitespace = true): string
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DomCrawler/Crawler.php]: That's a new feature. This part should be split in another PR for 6.3.

**Review by nicolas-grekas — COMMENTED** [review]
Branch 5.4 accepts only bug fixes. Please remove the new arguments from this PR and keep only the bugfix part. Then please submit another PR for 6.3 to consider adding the new arguments to 6.3.

File: src/Symfony/Component/DomCrawler/Crawler.php (hunk 1 above)
Comment (otsch) [reply src/Symfony/Component/DomCrawler/Crawler.php]: Thank you for the review. I removed it. 👍🏻

**Comment by otsch** [comment]
> Branch 5.4 accepts only bug fixes. Please remove the new arguments from this PR and keep only the bugfix part. Then please submit another PR for 6.3 to consider adding the new arguments to 6.3.

Thanks, changed it 👍🏻

**Comment by otsch** [comment]
@nicolas-grekas 
I now also fixed the code style issues from the fabbot.io github action. Sadly the "Tests 8.0" action failed now, but I think it's not related to the code change, it says "Docker pull failed". For the continuous integration action: I'm not sure if this one should work? Not sure if it fails because of the change?

**Comment by nicolas-grekas** [comment]
On a closer look, I'm not sure we should change this.
innerText is something that one should use on tags that don't have child nodes.
but I don't see how useful it'd be to rely on the head+tail being concatenated honeslty.
also, from a BC POV, this changes a behavior that others might rely on.
all in all, I'd be :-1: without stronger arguments to consider this a bug

**Comment by otsch** [comment]
> innerText is something that one should use on tags that don't have child nodes.

If that's the case I'm misunderstanding it. In the docs it says:

> innerText() is similar to text() but only returns the text that is the direct descendant of the current node, excluding any child nodes

So, for me this means i get the text from a node excluding the text from child nodes. The original test case also checks if the text from a child node is contained with `text()` but not with `innerText()`.

I understand questioning the usefulness of this method, as it can pretty easily lead to undesirable results when document structure changes.

But let's consider this example:
```php
$html = <<<HTML
    <div class="sun">
        <div class="sunrise">
            <span>▲</span> 07:49
        </div>
        <div class="sunset">
            <span>▼</span> 16:12
        </div>
    </div>
    HTML;

$dom = new \Symfony\Component\DomCrawler\Crawler($html);

var_dump($dom->filter('.sun .sunrise')->innerText());
```

The return value of `innerText()` is an empty string in this case.

The same example with the up and down arrows after the time:

```php
$html = <<<HTML
    <div class="sun">
        <div class="sunrise">
            07:49 <span>▲</span>
        </div>
        <div class="sunset">
            16:12 <span>▼</span>
        </div>
    </div>
    HTML;

$dom = new \Symfony\Component\DomCrawler\Crawler($html);

var_dump($dom->filter('.sun .sunrise')->innerText());
```

In this case `innerText()` returns `07:49`. 

Or consider table cells like this:
```html
<td><span class="currency">€</span> 120</td> <!-- innerText empty string -->
<td>120 <span class="currency">€</span></td> <!-- innerText 120 -->
```

In my understanding those are some real world use cases for this method and therefor I'd consider the current behavior a bug.

**Comment by nicolas-grekas** [comment]
Is that based on a real world use case that you have?

**Comment by otsch** [comment]
> Is that based on a real world use case that you have?

Yes, the first one with sunrise and sunset is. I tried to get the time using `innerText` and was surprised that it didn't work. I've already used the method before. Must have been the first time I tried to use it when the text comes after the child node.

**Comment by nicolas-grekas** [comment]
What about returning the tail only if the head is empty? That would cover the use case that you describe (and that makes sense to me), without trying to solve the "concat" use case, which doesn't to me.

**Comment by otsch** [comment]
> What about returning the tail only if the head is empty? That would cover the use case that you describe (and that makes sense to me), without trying to solve the "concat" use case, which doesn't to me.

@nicolas-grekas 👍🏻 sounds good to me, I also thought about that. I'll change it.

**Comment by otsch** [comment]
@nicolas-grekas done 🙂
It's now returning the value of the first (non empty) text child node. It doesn't concatenate.

**Comment by nicolas-grekas** [comment]
LGTM thanks. Can you please add a test case that has both head and tail, asserting that only head is returned?

**Comment by otsch** [comment]
@nicolas-grekas also done ☑️

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/DomCrawler/Crawler.php [hunk 2]
```diff
@@ -634,7 +634,13 @@ public function text(string $default = null, bool $normalizeWhitespace = true)
      */
     public function innerText(): string
     {
-        return $this->filterXPath('.//text()')->text();
+        foreach ($this->getNode(0)->childNodes as $childNode) {
+            if (\XML_TEXT_NODE === $childNode->nodeType && '' !== trim($childNode->nodeValue)) {
+                return $this->normalizeWhitespace($childNode->nodeValue);
```
Comment (fabpot) [inline src/Symfony/Component/DomCrawler/Crawler.php]: Would it make sense to have a flag as we have for `text()`?

Comment (nicolas-grekas) [reply src/Symfony/Component/DomCrawler/Crawler.php]: I wondered about this also. As is, the PR keeps the current behavior. Adding a flag could be useful, but nobody reported a use case so :shrug:

Comment (fabpot) [reply src/Symfony/Component/DomCrawler/Crawler.php]: But then again, it's not consistent with `text`

Comment (otsch) [reply src/Symfony/Component/DomCrawler/Crawler.php]: I added it in the first version I pushed, but I thought I should remove it @nicolas-grekas, because you said
> Please remove the new arguments from this PR and keep only the bugfix part.

When it's OK as an optional argument `bool $normalizeWhitespace = true` I can add it again.

Comment (otsch) [reply src/Symfony/Component/DomCrawler/Crawler.php]: OK? https://github.com/symfony/symfony/pull/48684/commits/48d39856a6854f8b529879dd0a216044c545fb53

File: src/Symfony/Component/DomCrawler/Tests/AbstractCrawlerTest.php [hunk 3]
```diff
@@ -361,9 +361,19 @@ public function testText()
     public function testInnerText()
     {
         self::assertCount(1, $crawler = $this->createTestCrawler()->filterXPath('//*[@id="complex-element"]'));
-
```
Comment (fabpot) [inline src/Symfony/Component/DomCrawler/Tests/AbstractCrawlerTest.php:364]: Shoud be reverted

File: src/Symfony/Component/DomCrawler/Crawler.php [hunk 4]
```diff
@@ -632,9 +632,15 @@ public function text(string $default = null, bool $normalizeWhitespace = true)
     /**
      * Returns only the inner text that is the direct descendent of the current node, excluding any child nodes.
      */
-    public function innerText(): string
+    public function innerText(bool $normalizeWhitespace = true): string
```
Comment (fabpot) [inline src/Symfony/Component/DomCrawler/Crawler.php:565]: Adding an argument is a BC break and should be done a bit differently:

```suggestion
    public function innerText(/* bool $normalizeWhitespace = true */): string
```

Then, add a phpdoc to document the new argument:

```
/**
 * @param bool $normalizeWhitespace
 */
```

Finally, use `func_num_args` and `func_get_arg` to retrieve the argument:

```
$normalizeWhitespace = 1 <= \func_num_args() ? func_get_arg(0) : false;
```

Note that I'm using `false` by default to keep the current behavior in 6.3.

Last, if someone does not pass the argument, raise a deprecation warning that the default value in 7.0 will change to `true`. So, if needed, people need to explicitly pass `false` to keep the current behavior in 7.0.

**Review by fabpot — CHANGES_REQUESTED** [review]
As this is a new behavior, this PR should target 6.3 instead of 5.4.

File: src/Symfony/Component/DomCrawler/Crawler.php (hunk 4 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DomCrawler/Crawler.php:565]: About the default: the previous code was doing `$this->filterXPath('.//text()')->text()`, where `->text()` is the very method on this class, which defaults to `$normalizeWhitespace = true`, so it's fine using `true` as default for this method also.

Comment (otsch) [reply src/Symfony/Component/DomCrawler/Crawler.php:565]: Thanks. I also just wanted to comment that. I think the new argument just offers the option to set it to `false` which isn't breaking backwards compatibility I guess.

**Comment by otsch** [comment]
> As this is a new behavior, this PR should target 6.3 instead of 5.4.

@fabpot Is this meant independently of your comment about the new `$normalizeWhitespace` argument?

File: src/Symfony/Component/DomCrawler/Crawler.php (hunk 4 above)
Comment (fabpot) [reply src/Symfony/Component/DomCrawler/Crawler.php:565]: @nicolas-grekas You're right
@otsch Still for 6.3 anyway ;)

**Comment by OskarStark** [comment]
Closing in favor of https://github.com/symfony/symfony/pull/48940

**Comment by otsch** [comment]
@fabpot @nicolas-grekas I've created a new PR with target branch 6.3 as requested: https://github.com/symfony/symfony/pull/48940 
(if the target branch can be changed in an existing PR, I haven't found how it can be done)

**Comment by xabbuh** [comment]
@otsch it's hidden behind the "edit" button: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request

**Comment by otsch** [comment]
> @otsch it's hidden behind the "edit" button: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request

@xabbuh Oh, looks like I'm blind 😂🙈 
For me it always looked like only the title changes to an Input when you click Edit. Didn't notice that also the target branch changes to a select. Thx for the hint!
